File: src/main/java/com/beta_foprhoton/creativecomputerbugs/registy/BugComputerHolderRegister.java
package com.beta_foprhoton.creativecomputerbugs.registy;

import com.beta_foprhoton.creativecomputerbugs.foundation.computercraft.core.BugComputerHolder;
import dan200.computercraft.api.pocket.IPocketUpgrade;
import dan200.computercraft.api.upgrades.UpgradeData;
import dan200.computercraft.shared.computer.core.ComputerFamily;
import net.minecraft.core.Direction;
import net.minecraft.world.level.block.entity.BlockEntity;

import java.util.HashMap;
import java.util.Map;

import static net.minecraft.world.level.block.Block.popResource;

public class BugComputerHolderRegister {
    private final Map<Integer , BugComputerHolder> registry;
    private int idCounter = 0;

    public BugComputerHolderRegister () {
        this.registry = new HashMap<>();
    }

    public BugComputerHolder getHolder (int ID) {
        return registry.get(ID);
    }
    
    public void tick () {
        registry.forEach((id, holder) -> {
            holder.serverTick();
        });
    }

    public BugComputerHolder create (BlockEntity infectBlockEntity, UpgradeData<IPocketUpgrade> upgrade, ComputerFamily family, Direction direction) {
        var holderOld = BugComputerHolder.getBugComputerHolder(infectBlockEntity);
        if (holderOld != null) return null;
        var holder = new BugComputerHolder(infectBlockEntity, CCBItems.BUG_WORM.asStack(), upgrade, family, direction, idCounter);
        registry.put(holder.id, holder);
        idCounter ++;
        return holder;
    }

    public void remove (int ID) {
        var holder = registry.get(ID);
        holder.unload();
        registry.remove(ID);
        popResource(holder.getLevel(), holder.getInfectedBlockEntity().getBlockPos(), holder.getBugItem());
    }
}

File: src/main/java/com/beta_foprhoton/creativecomputerbugs/registy/BugComputerHolderRegister.kt
package com.beta_foprhoton.creativecomputerbugs.registy;

import com.beta_foprhoton.creativecomputerbugs.foundation.computercraft.core.AbstractBugComputerHolder
import com.beta_foprhoton.creativecomputerbugs.foundation.computercraft.core.block.WormComputerHolder
import com.beta_foprhoton.creativecomputerbugs.foundation.computercraft.core.entity.ParasiteComputerHolder
import com.beta_foprhoton.creativecomputerbugs.foundation.helpers.extensions.BlockEntityExtensions.getBugComputerHolder
import com.beta_foprhoton.creativecomputerbugs.foundation.helpers.extensions.EntityExtensions.getBugComputerHolder
import dan200.computercraft.api.pocket.IPocketUpgrade
import dan200.computercraft.api.upgrades.UpgradeData
import dan200.computercraft.shared.computer.core.ComputerFamily
import net.minecraft.core.Direction
import net.minecraft.world.entity.Entity
import net.minecraft.world.level.block.entity.BlockEntity

class BugComputerHolderRegister(private var idCounter: Int) {
    private val wormRegistry = HashMap<Int , WormComputerHolder>()
    private val parasiteRegistry = HashMap<Int , ParasiteComputerHolder>()

    fun getWormComputerHolder(id: Int): WormComputerHolder? {
        return wormRegistry[id]
    }

    fun getParasiteComputerHolder(id: Int): ParasiteComputerHolder? {
        return parasiteRegistry[id]
    }
    
    fun tick() {
        wormRegistry.forEach { it.value.tick() }
        parasiteRegistry.forEach { it.value.tick() }
    }

    fun createWormComputerHolder(infectBlockEntity: BlockEntity, upgrade: UpgradeData<IPocketUpgrade>?, family: ComputerFamily): WormComputerHolder? {
        val holderOld = infectBlockEntity.getBugComputerHolder()
        if (holderOld != null) return null
        val holder = WormComputerHolder(
            family = family,
            bugItem = CCBItems.BUG_WORM.asStack(),
            upgrade = upgrade,
            id = idCounter,
            blockEntity = infectBlockEntity
        )
        wormRegistry[holder.id] = holder
        idCounter ++
        return holder
    }

    fun createParasiteComputerHolder(entity: Entity, upgrade: UpgradeData<IPocketUpgrade>?, family: ComputerFamily): ParasiteComputerHolder? {
        val holderOld = entity.getBugComputerHolder()
        if (holderOld != null) return null
        val holder = ParasiteComputerHolder(
            family = family,
            bugItem = CCBItems.BUG_WORM.asStack(),
            upgrade = upgrade,
            id = idCounter,
            entity = entity
        )
        parasiteRegistry[holder.id] = holder
        idCounter ++
        return holder
    }

    fun remove(id: Int) {
        val holder1 = wormRegistry[id]
        if (holder1 != null) {
            holder1.unload()
            wormRegistry.remove(id)
            holder1.popResource()
        }
        val holder2 = parasiteRegistry[id]
        if (holder2 != null) {
            holder2.unload()
            parasiteRegistry.remove(id)
            holder2.popResource()
        }
    }
}